application-xpoll-api/src/main/java/com/xwiki/xpoll/internal/CondorcetResultsCalculator.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package com.xwiki.xpoll.internal;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import com.xpn.xwiki.objects.BaseObject;
+
+/**
+ * Calculates the result of a poll in a condorcet way. The Schulze algorithm is used to rank the entries of the poll.
+ * The winners of the poll are the entries satisfy the relation: p[E,X] >= p[X,E] where E is the winning entry, X any
+ * other entry and p[A,B] equals to the number of voters who prefer the entry A over B.
+ *
+ * @version $Id$
+ * @since 2.1
+ */
+public class CondorcetResultsCalculator implements PollResultsCalculator
+{
+    @Override
+    public Map<String, Integer> getResults(List<BaseObject> xpollVotes,
+        List<String> proposals)
+    {
+        Map<String, Map<String, Integer>> ballotsMatrix = new HashMap<>();
+
+        for (String proposal : proposals) {
+            ballotsMatrix.put(proposal, new HashMap<>());
+            for (String s : proposals) {
+                ballotsMatrix.get(proposal).put(s, 0);
+            }
+        }
+
+        computeProposalWeights(xpollVotes, ballotsMatrix, proposals);
+
+        Map<String, Map<String, Integer>> p = new HashMap<>();
+        for (String proposal : proposals) {
+            p.put(proposal, new HashMap<>());
+            for (String s : proposals) {
+                p.get(proposal).put(s, 0);
+            }
+        }
+
+        computeProposalStronghestPaths(proposals, ballotsMatrix, p);
+
+        Map<String, Integer> scores = new HashMap<>();
+        // E wins since p[E,X] >= p[X,E] for every other candidate X.
+        for (String p1 : p.keySet()) {
```

Comment:
```suggestion
        for (String proposal : proposals) {
```

Isn't ``p.keySet()`` the list of proposals?

Reply:
yes, it is. But I thought it would be clearer if I iterated over the keyset as if I were iterating over a matrix, rather than iterate twice over the proposals.